### PR TITLE
Add a Docker container for doing automated builds for CloudABI.

### PR DESCRIPTION
**Pull request by EdSchouten**
Setting up a cross compilation toolchain for CloudABI is relatively
easy. It's just a matter of installing a somewhat recent version of
Clang (5.0 preferred) and installing the corresponding
`${target}-cxx-runtime` package, containing a set of core C/C++ libraries
(libc, libc++, libunwind, etc).

Eventually it would be nice if we could also run `x.py test`. That,
however still requires some more work. Both libtest and compiletest
would need to be adjusted to deal with CloudABI's requirement of having
all of an application's dependencies injected. Let's settle for just
doing `x.py dist` for now.

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -126,6 +126,8 @@ matrix:
       if: branch = auto
     - env: IMAGE=dist-armv7-linux DEPLOY=1
       if: branch = auto
+    - env: IMAGE=dist-cloudabi DEPLOY=1
+      if: branch = auto
```
Comment (kennytm) [inline .travis.yml]: I don't think we want to put this as a separate job on the CI for now due to resource limitation.

**Review by kennytm — COMMENTED** [review]
Given that we are not going to add a new CI job in the short term, could you move the `dist-cloudabi` folder as `src/ci/docker/disabled/dist-cloudabi`?

**Comment by EdSchouten** [comment]
Aww... That's too bad. Be sure to get in touch with me if CPU capacity opens up.

**Comment by EdSchouten** [comment]
Sorry if I'm sounding a bit negative, but not having automated builds for CloudABI is going to be more annoying than I had anticipated. If no automated builds can be provided, my idea would be to set up an automated build myself. That way, people can install Rust nightly on their operating system of choice and extend that with CloudABI support by extracting a `rust-std-nightly-x86_64-unknown-cloudabi.tar.xz` tarball hosted by me.

It turns out this won't work in practice, because `rustc` is quite picky:

```
error[E0514]: found crate `core` compiled by an incompatible version of rustc
  |
  = help: please recompile that crate using this compiler (rustc 1.25.0-nightly (e6072a7b3 2018-01-13))
  = note: the following crate versions were found:
          crate `core` compiled by rustc 1.25.0-dev (3f92e8d89 2018-01-14): /home/ed/.rustup/toolchains/nightly-x86_64-unknown-freebsd/lib/rustlib/x86_64-unknown-cloudabi/lib/libcore-14da3e8a93d4bc72.rlib
```

Do you by any chance know what would be the most elegant way to deal with this?

**Comment by Mark-Simulacrum** [comment]
A potential solution is https://github.com/japaric/xargo. However, what you suggest would work, you'd just have to ship all of the "std" component that we normally do, including `core`. In theory, I could see rustup being extended to support using "components" from sources other than static.rust-lang.org, but I don't think there's a current maintainer of rustup to talk to about that.

**Comment by malbarbo** [comment]
@EdSchouten You can try adding this to dist-various-1 or dist-various-2 builder.

**Comment by EdSchouten** [comment]
@kennytm Integrating it in `dist-various-*` would be fine for me. Would that be all right for you?

**Comment by kennytm** [comment]
@EdSchouten `dist-various-*` uses Ubuntu 16.04 LTS. Does this work for CloudABI? I'm not sure if we want to upgrade them to 17.10.

**Comment by alexcrichton** [comment]
@EdSchouten ah yes @Mark-Simulacrum and @malbarbo are both right in that if this is a separate builder we don't currently have the capacity to add more so it'll need to be `disabled` for now. If, however, it can be folded into an existing builder (which as it's just libstd I believe it should be totally feasible to do this) then our existing builders should have enough capacity to cross compile a new platform. We typically just pick whichever is fastest and throw it in there. Looks like the [last successful run](https://travis-ci.org/rust-lang/rust/builds/329360430) had `dist-various-1` clocking in at 1h20m and `dist-various-2` clocking in at 1h9m, so adding to `dist-various-2` should do the trick!

Most of our containers use 16.04 as that was the latest when they were created, but I'm all for updating them as eagerly as possible (pending we're able to fix accidental regressions that creates of course). The `dist-various-*` containers should be able to get updated any time, the build will fail if they fail to build! Otherwise we in general don't want to have a dependency on a specific version, so if we are accidentally relying on something 16.04-specific we can fix that later.

In other words, let's see if we can fold this into `dist-various-2`? And you can feel free to update dist-various-2 if you'd like. You'll want to run locally though to make sure it all still works via `./src/ci/docker/run.sh dist-various-2`

**Comment by EdSchouten** [comment]
Hi @alexcrichton. Thanks for clarifying!

I just integrated CloudABI support into `dist-various-2`. It worked all right, but I did have to set `DEPLOY=1`. Apparently that's needed to work around #46781 / #43982. Let me know what you think!

**Comment by alexcrichton** [comment]
Thanks @EdSchouten! I wonder if we could perhaps start with 1 or 2 targets just to make sure things work? We can always expand if we've still got capacity afterwards!

**Comment by EdSchouten** [comment]
Done! \o/

**Comment by EdSchouten** [comment]
@bors retry

Hmm... odd. The Travis page mentioned this build got cancelled? Not sure whether I have rights to force a retry.

**Comment by kennytm** [comment]
@EdSchouten The build is canceled because Travis ded https://www.traviscistatus.com/incidents/c1s4dlyxd4lf

The build hasn't actually started, so `retry` does nothing (besides insufficient privilege).
